### ISSUE TITLE
Building a project between two different SDKs fails hard if you reuse nodes

### ISSUE THREAD

**Issue by davkean** [dotnet/sdk#2161]
## Steps to reproduce

1. Build a project via msbuild with 2.1.300-preview2-008324
2. Before msbuild nodes shutdown, build a project via msbuild with 2.1.300-preview2-008533

## Expected  behavior
No errors

## Actual behavior
Actual:
```
C:\Program Files\dotnet\sdk\2.1.300-preview2-008533\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInfer
ence.targets(134,5): error MSB4062: The "ShowPreviewMessage" task could not be loaded from the assembly C:\Program File
s\dotnet\sdk\2.1.300-preview2-008533\Sdks\Microsoft.NET.Sdk\targets\..\tools\net46/Microsoft.NET.Build.Tasks.dll.  Conf
irm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the
 task contains a public class that implements Microsoft.Build.Framework.ITask. [E:\project-system2\src\Microsoft.Visual
Studio.ProjectSystem.VisualBasic.UnitTests\Microsoft.VisualStudio.ProjectSystem.VisualBasic.UnitTests.csproj]
```

## Environment data
`dotnet --info` output:

**Comment by davkean** [comment]
Elsewhere we version tasks dlls every build to avoid this.

**Comment by livarcocc** [comment]
@nguerrera and I were just talking about this. We used to version assemblies out of the SDK for every build. It seems like we regressed with the infra-changes we did moving to repotoolset.

@tannergooding can you investigate this. I will mark this for 2.1.3xx and will consider taking it to shiproom.

**Comment by nguerrera** [comment]
Yes, this is a regression. Task dlls should get a unique assembly version with every build.

**Comment by nguerrera** [comment]
Original issue: #488, original fix #815

**Comment by tannergooding** [comment]
PR is here: https://github.com/dotnet/sdk/pull/2163

**Comment by nguerrera** [comment]
I think this regressed again with a repo toolset update :(

**Comment by davkean** [comment]
Ugh, how do we prevent this?

**Comment by nguerrera** [comment]
I'll

**Comment by nguerrera** [comment]
I'll add a test that AssemblyVersion == FileVersion and neither have a fourth part of 0. I think it will only ever fail in official builds due to how we go to 42.42.42.42 otherwise in repo toolset, but that's a lot better than nothing.

Interestingly, though, the reason I even checked was an issue on .net core msbuild and the conflict was between different minor versions. It appears that core might not allow different versions. I will double check and file an issue on msbuild. I think we never noticed this because core didn't have node reuse until recently.

**Comment by JamesNK** [comment]
What's the status of this bug? Customers are encountering problems with the build tasks in Grpc.Tools:

https://github.com/grpc/grpc/issues/20501
